src/fetch-or-populate-file-cache.ts

```diff
@@ -0,0 +1,77 @@
+import { fileExists, readJsonFile, writeJsonFile } from '@metamask/utils/node';
+import type { Json } from '@metamask/utils/node';
+
+import { ONE_HOUR } from './constants';
+import { createModuleLogger, projectLogger } from './logging-utils';
+
+const log = createModuleLogger(projectLogger, 'fetch-or-populate-file-cache');
+
+/**
+ * The data stored in the cache file.
+ */
+type FileCache<Data extends Json> = {
+  /**
+   * When the data was stored.
+   */
+  ctime: string;
+  /**
+   * The cached data.
+   */
+  data: Data;
+};
+
+/**
+ * How long to cache data retrieved from an API (to prevent rate limiting).
+ *
+ * Equal to 1 hour.
+ */
+const DEFAULT_MAX_AGE = ONE_HOUR;
+
+/**
+ * Avoids rate limits when making requests to an API by consulting a file cache.
+ *
+ * Reads the given cache file and returns the data within it if it exists and is
+ * fresh enough; otherwise runs the given function and saves its return value to
+ * the file.
+ *
+ * @param args - The arguments to this function.
+ * @param args.filePath - The path to the file where the data should be saved.
+ * @param args.getDataToCache - A function to get the data that should be cached
+ * if the cache does not exist or is stale.
+ * @param args.maxAge - The amount of time (in milliseconds) that the cache is
+ * considered "fresh". Affects subsequent calls: if `fetchOrPopulateFileCache`
+ * is called again with the same file path within the duration specified here,
+ * `getDataToCache` will not get called again, otherwise it will. Defaults to 1
+ * hour.
+ */
+export async function fetchOrPopulateFileCache<Data extends Json>({
+  filePath,
+  maxAge = DEFAULT_MAX_AGE,
+  getDataToCache,
+}: {
+  filePath: string;
+  maxAge?: number;
+  getDataToCache: () => Data | Promise<Data>;
+}): Promise<Data> {
+  const now = new Date();
+
+  if (await fileExists(filePath)) {
+    const cache = await readJsonFile<FileCache<Data>>(filePath);
```

Comment:
Nit: Have you considered trying to read the file directly and handling the "does not exist" error, rather than checking for existence first? That would align better with [what the Node.js docs recommend](https://nodejs.org/api/fs.html#fsexistspath-callback):

>Using fs.exists() to check for the existence of a file before calling fs.open(), fs.readFile(), or fs.writeFile() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file does not exist.